eos-paygd/main.c

```diff
@@ -194,24 +239,35 @@ main (int   argc,
       watchdog_id = g_timeout_add_seconds_full (G_PRIORITY_HIGH, 10, ping_watchdog, NULL, NULL);
       g_assert (watchdog_id > 0);
 
-      /* Let systemd know it's okay to proceed to pivot to the final root */
-      sd_notify_ret = sd_notify (0, "READY=1");
+      /* Here be dragons:
+       * We're currently in the initramfs root directory, and systemd is putting
+       * all the useful bits of the system into /sysroot in preparation for the
+       * root pivot. After the root pivot, it will delete everything in the
+       * initramfs root, which will leave us in limbo forever.  We need to
+       * chroot() out of this into sysroot.
+       *
+       * However, once we do that, we no longer have access to the socket we
+       * need to write to to inform systemd we're ready for it to do the root
+       * pivot.
+       *
+       * What we do is chdir() into the directory $NOTIFY_SOCKET resides in,
+       * chroot() into /sysroot, then use our own variant of sd_notify() that
+       * doesn't refuse to use relative paths to notify systemd we're ready.
+       *
+       * After that, we can chdir() into our new / and begin trying to connect
+       * to the dbus socket. */
+
+      sd_socket_env = g_getenv("NOTIFY_SOCKET");
+      sd_socket_dir = g_path_get_dirname(sd_socket_env);
+      sd_socket_name = g_path_get_basename(sd_socket_env);
+      chdir(sd_socket_dir);
+      chroot("/sysroot");
```

Comment:
Is `/sysroot` guaranteed to exist yet? My understanding is that gets created at some point while the initrd is running but is it guaranteed to be before `eos-payg-2.service` starts? We only have `After=sysinit.target` but I think we need `After=sysroot.mount` if I'm reading this correctly: https://www.freedesktop.org/software/systemd/man/bootup.html#

Reply:
The /sysroot mount point is actually a dir in the dracut image, so it's technically always there... and since we open the dbus socket with an absolute path, that will work.  It's probably a good idea to chdir("/"); yet one more time after we successfully connect to dbus, so relative paths will work - we'll probably be "under" the root mount for relative paths until I add  that..

Reply:
Newly pushed changes add the chdir("/") after dbus connect